### PR TITLE
Run4-hgx199 Correct bug for HFNose and extend validity check for silicon cells

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Correct bug for HFNose which made holes in layer 8. Also for all silicon wafers, the validity of DetId's is extended. Replace some of the cost statements with LogVerbatim

#### PR validation:

Tested with all the relevant Phase2 workflows

#### if this PR is a backport please specify the original PR:

No back porting is needed

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@ianna @cvuosalo @civanch @kpedro88 Could you approve this?

**Comment by kpedro88** [comment]
+upgrade

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 1]
```diff
@@ -495,17 +496,25 @@ bool HGCalDDDConstants::isValidHex8(int layer, int modU, int modV, int cellU, in
...
-  }
+  if (((cellV - cellU) >= N) || ((cellU - cellV) > N))
+    return false;
+
+  auto ktr = hgpar_->waferTypes_.find(indx);
+  if (ktr == hgpar_->waferTypes_.end())
+    return true;
+
+  //  edm::LogVerbatim("HGCalGeom") << "Corners " << (ktr->second).first << ":" << waferVirtual(layer,modU,modV);
+  int type = ((itr == hgpar_->typesInLayers_.end()) ? 2 : hgpar_->waferTypeL_[itr->second]);
```
Comment (cvuosalo) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: Does `type == 2` have some meaning? Should 2 be a constant? It looks like `type` is boolean. It is checked for `type == 0` in `isValidCell8()`, so maybe its values should be 0 or 1, but not 2? Or will `type` later take on a wider range of values?

Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: Type could be 0, 1 or 2 for silicon wafers at least in this context. Type is one attribute of silicon cell.

Comment (cvuosalo) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: Is it worth having an `enum` for the three silicon cell types? Will silicon cell type be referenced elsewhere in the code?
If not, a short comment would be helpful here:
`// Silicon wafers can be type 0, 1, or 2.`

Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: I shall introduce enum later. Please sign this one. This PR solves an issue for HFNose detector. Need this one to be integrated fast

**Comment by cvuosalo** [comment]
+1

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc (hunk 1 above)
Comment (cvuosalo) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: Sounds good.

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
@bsunanda I see that this PR is impacting on scenario D41, used for L1T TDR. Could you please comment about the expected effect of this fix?

**Comment by bsunanda** [comment]
@fabiocos Yes it does. These are bug fixes. Should I try to backport some of the bug fixes to 10_6_X? But it is important that this PR to be integrated first. There is also an issue of Hcal which may need back porting for Phase2 trigger studies.

**Comment by fabiocos** [comment]
@bsunanda for the time being I will integrate this PR in master, then we will discuss with PPD what are the plans for activities in 10_6_X

**Comment by rekovic** [comment]
@bsunanda (attn @kpedro88)
>@fabiocos Yes it does. These are bug fixes. Should I try to backport some of the bug fixes to 10_6_X? But it is important that this PR to be integrated first. There is also an issue of Hcal which may need back porting for Phase2 trigger studies.
 
What effect and how large is this PR expected to make for Phase2?  
Can you also clarify the statement about the Hcal issues? 
Where are these changes relevant , GEN / DIGI ?

**Comment by bsunanda** [comment]
For HGCal it affects DIGI and beyond. For HCal the issue will affect only the trigger part

**Comment by bsunanda** [comment]
@fabiocos  @rekovic It is not clear to me in the meeting if this needs back porting to 10_6

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
@bsunanda @rekovic we did not take any final decision. I believe we need an assessment of the consequences of this issue on D41, and in any case the use case for a backport should be clarified, as the production is basically already made

**Comment by rekovic** [comment]
@bsunanda 
> For HGCal it affects DIGI and beyond.

So this means that the DIGIs made in 10_6_0_patch2  will not be readable with this PR ?  I.e. we cannot re-make HGC TPs from HGC DIGIs made before ?

**Pull request merged**